Document why, how, and what for RuboCop::Packaging

utkarsh2102 opened the pull request:
This PR covers most of the reasons why the `Packaging` extension was born and explains the cops that are provided by the `Packaging` extension, describing the need to write and enforce it and how is it problematic on the Debian (downstream) side.

Resloves and relates to: https://github.com/utkarsh2102/rubocop-packaging/issues/3

Signed-off-by: Utkarsh Gupta <<utkarsh@debian.org>>

utkarsh2102 committed:
Init with introduction

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Add other starting headers

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Add about this tool and GemspecGit cop

More verbosely, add:
- Why Packaging Extension?
- Cops Description
- Gemspec Git
  - Rationale
  - Need
  - Examples

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Extend description for RelativeRequireToLib cop

More verbosely, add:
- Relative Require To Lib
  - Rationale
  - Need
  - Examples

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Re-order the examples

Until now, we were mentioning all the bad ones together
and all the good ones together!

However, it appears to be much more neater and readable
if we alternate b/w the bad one and the good one.
This way, we can clearly express the bad part in the bad
example and how could it be tweaked to make it "good"!

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Proof-read the entire document

- Fix typos.
- Fix grammar.

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

pravi commented on README.adoc:
```diff
@@ -0,0 +1,223 @@
+
+==== Need [[gemspec-git-need]]
+
+Packages in Debian are built in a clean environment (https://wiki.debian.org/sbuild[sbuild],
```
I think you can also mention the source package consists of released tar ball (usually downloaded from github tags/releases page or converted from the .gem file), which is extracted during build. So even if we add git as a build dependency, it would still fails as the debian package source tree is not a git repo. Even when the package is maintained in git, it is uploaded as tarballs to the archive without any version control information.

utkarsh2102 replied:
This is a fantastic point! :heart: 
Adding.. \o/

terceiro replied:
+1

pravi commented on README.adoc:
```diff
@@ -0,0 +1,223 @@
+for instance:
+
+* `ruby-core` as part of their CI system runs their test suite against an unpackaged Ruby
+  tarball which doesn't have a `.git` folders. That means they needed to overwrite the
```
a `.git` folder (singular) - or may be use directory instead of folder as folder is usually referred in graphical environments I think.

utkarsh2102 replied:
Great catch, fixed!

pravi commented on README.adoc:
```diff
@@ -0,0 +1,223 @@
+Originally, `git ls-files` inside the default gemspec template was designed so that users
+publishing their first gem wouldn't unintentionally publish artifacts to it. +
+Now that `bundler` won't let you release if you have uncommitted stuff in your working
+repository, the risk is quite lower (if none at all!).
```
(if none at all!) => probably rephrase it

utkarsh2102 replied:
You're right, it needs rephrasing. Thanks!

terceiro replied:
"Recent versions of bundler won't let you release if you have uncommitted files in your working directory, so that risk is lower." ?

However this does not solve the concern of excluding editor backup files and othet temporary files in the release

utkarsh2102 replied:
Added!
However, didn't add the last line...yet.

pravi commented on README.adoc:
```diff
@@ -0,0 +1,223 @@
+
+# good
+require_relative 'spec_helper'
+require_relative 'foo/bar'
```
require 'foo/bar' I assume

utkarsh2102 replied:
I wanted to express that `require_relative` is okay to use inside `lib/` or inside `specs/`, as long as it is not cross spec-lib.
So I'll perhaps amend it to `require_relative 'spec/foo/bar'`.

utkarsh2102 committed:
Add Contributing section

And also add a "How to Contribute?" subsection.

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Add License \o/

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Add "Spread the Word" section

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Add Credits section

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

terceiro commented on README.adoc:
```diff
@@ -0,0 +1,275 @@
+we also run the tests with the `lib/` directory removed to force those tests to use the
+library installed globally via the Debian package. +
+We do so because we want to ensure that the library is loaded from the system location and not
+from the source tree and that it works like it is intended to.
```
I think the explanation for this can be improved. What about this:

Debian has a [testing infrastructure](https://ci.debian.net/) that is designed to test packages in their installed form, i.e. closer to how an end used would use it than to how a developer work against it. For this to work, the test suite must load code that's installed system-wide, instead of the code in the source tree. Using `require_relative` from the tests into the `lib` directory makes that impossible, but it also makes the test look less like client code that would use the code in your gem. Therefore, we recommend that test code uses the main library code without `require_relative`.

utkarsh2102 replied:
Nicely phrased, added! :+1:

terceiro commented on README.adoc:
```diff
@@ -0,0 +1,275 @@
+Of course, that said, with this cop, we don't intend to disregard the usage of `require_relative`
+in general. Not at all! It is still recommended but we just add one tiny exception to it. +
+That is, avoid using `require_relative` with a relative path to `lib/` in your tests. That's
+it, that's all we ask!
```
I would say it like this:

We want to emphasize that **there is nothing wrong** with using `require_relative` inside `lib/`, it's just using it from your test code to the `lib` directory prevents the "test the library installed system-wide" use case.

utkarsh2102 replied:
Great, added! :+1: 
But also kept the last conclusive line that we'd still recommend to use `require_relative` with just this one exception.

terceiro commented on README.adoc:
```diff
@@ -0,0 +1,275 @@
+Listed below are the cops provided by the `Packaging` extension, describing the need to write
+and enforce it and how is it problematic on the Debian (downstream) side.
+
+=== Gemspec Git [[gemspec-git]]
```
Instead of focusing on the cop names, I would instead use names that describe the actual antipatterns we are trying to get people to avoid, e.g. "git usage in gemspec", "require_relative from test code into lib". You can still mention the cop name, but that would be a detail instead of the name of the section.

utkarsh2102 replied:
Hm, okay..
I've tried to tackle this in the latest commit: https://github.com/utkarsh2102/packaging-style-guide/pull/1/commits/ef66175dd69acdbb055c907e38d18058caa79add

utkarsh2102 committed:
Paraphrase some lines

Thanks to Praveen's and Antonio's amazing
suggestions! 🔥

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 committed:
Describe the actual antipatterns with cop names

Based on Antonio's suggestion that it's better to
use names that describe the actual antipatterns that
we're trying to get people to avoid.

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 commented:
Thanks, @pravi and @terceiro, I've tried to include all the suggestions in the best possible way!
Please give it one final review and let me know if you find some scope of improvement! :shipit: 
(just know that in case of minor issues, we can always do some quick follow-up commits, I've just inclined to get this merged in the first place!)

pravi commented:
Sounds good to me.

utkarsh2102 committed:
Professionalize the guide

Don't use new lines. Change paragraphs wherever needed!
Thanks, Antonio \o/

Signed-off-by: Utkarsh Gupta <utkarsh@debian.org>

utkarsh2102 commented:
Well, okay. I am merging this for now..
In case you've got any more things, please raise an issue (or a PR :laughing:)!
